src/inputs/pcap/PcapInputStream.h

```diff
@@ -158,6 +176,33 @@ class PcapInputEventProxy: public visor::InputEventProxy
         pcap_stats_signal(stats);
     }
 
+    void register_matcher(const std::string &key, const std::string &schema)
+    {
+        std::unique_lock lock(_pcap_proxy_mutex);
+        if (_registered_matchers.count(key) != 0) {
+            throw PcapException(fmt::format("matcher with key [{}] is already registered", key));
+        }
+        _registered_matchers[key] = schema;
+    }
+
+    void register_udp_predicate_signal(const std::string &schema, const std::string &key, const std::string &value, UdpSignal callback)
+    {
+        std::unique_lock lock(_pcap_proxy_mutex);
+        if (const auto it = _registered_matchers.find(key); it == _registered_matchers.end()) {
+            throw PcapException(fmt::format("there is no matcher registered with key [{}]", key));
+        } else if (it->second != schema) {
+            throw PcapException(fmt::format("matcher schema [{}] does not match predicate schema [{}], they must be the same.", it->second, schema));
+        }
+        auto hash = key + value;
+        predicate_udp_signals[hash].push_back(callback);
+    }
+
+    void send_metadata(const std::string &key, const std::string &value)
```

Comment:
when is this called? your notes say in the udp_signal but i don't see it

Reply:
I've not added that call yet. but it would be in the DnsStreamHandler.cpp. Just added the overall idea.

Hmm I think we could use metadata.empty() and metada.clear() to remove any_match. I think it will not be a big change in the performance